### PR TITLE
[test] Add ert roofline check

### PR DESCRIPTION
**Pull request by jgphpc**
Will fix UES-213

### CODE REVIEW

File: cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        self.build_system.ldflags = ['-O3 -fopenmp']
+        self.sourcesdir = os.path.join(self.current_system.resourcesdir,
+                                       'roofline', 'ert')
+        self.rpt = '%s.rpt' % self.executable
+        self.maintainers = ['JG']
+        self.tags = {'scs'}
+
+    def setup(self, partition, environ, **job_opts):
+        super().setup(partition, environ, **job_opts)
+        # self.job.launcher.options = ['--cpu-bind=verbose,none']
```
Comment (jgphpc) [inline cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py]: todo

File: cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                    r'(?P<DRAMbw>\d+.\d+) DRAM EMP', self.roofline_rpt,
+                    'DRAMbw', float), DRAMbw, -0.1, 0.3),
+            ])
+        else:
+            self.post_run = [
+                'cat *_job.out | python2 preprocess.py > pre',
+                'python2 maximum.py < pre > max',
+                'python2 summary.py < max > sum',
+            ]
+            self.sanity_patterns = sn.assert_found('GFLOPs', 'sum')
```
Comment (jgphpc) [inline cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py]: add line

File: cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+        if not mpitask == 36:
```
Comment (teojgo) [inline cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py]: This should be added in the `setup` function. Also change to `if mpitask != 36`

File: cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        self.build_system.cppflags = [
+            '-DERT_ALIGN=32',
+            '-DERT_MEMORY_MAX=1073741824',
+            '-DERT_MPI=True',
+            '-DERT_OPENMP=True',
+            '-DERT_TRIALS_MIN=1',
+            '-DERT_WORKING_SET_MIN=1',
+            '-DERT_FLOP=%s' % flop,
+        ]
+        self.name = 'ert_FLOPS.' + '{:03d}'.format(flop) + \
```
Comment (teojgo) [inline cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py]: You can replace with:
```
self.name = 'ert_FLOPS.{:04d}_MPI.{:02d}_OpenMP.{:03d}'.format(
    flop, mpitask, ompthread)
```

**Comment by jgphpc** [comment]
Failed because of kesch: (?)
```
PrgEnv-cray(3):ERROR:105: 
Unable to locate a modulefile for 'PrgEnv-cray/1.0.2_gdr'
```

File: cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    The Empirical Roofline Tool, ERT, automatically generates roofline data.
+    https://bitbucket.org/berkeleylab/cs-roofline-toolkit/
+    """
+    def __init__(self):
+        super().__init__()
+        self.descr = 'Empirical Roofline Toolkit'
+        self.sourcesdir = os.path.join(self.current_system.resourcesdir,
+                                       'roofline', 'ert')
+        self.build_system = 'SingleSource'
+        self.sourcepath = 'kernel1.c driver1.c'
```
Comment (vkarak) [inline cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py]: Haha, good trick!

File: cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        self.executable = 'ert.exe'
+        self.build_system.ldflags = ['-O3 -fopenmp']
```
Comment (vkarak) [inline cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py]:
```suggestion
        self.build_system.ldflags = ['-O3', '-fopenmp']
```

File: cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        if self.num_tasks != 36:
+            self.job.launcher.options = ['--cpu-bind=verbose,none']
+
+
+@rfm.parameterized_test(
+    *[[mpitask, flop]
+    for mpitask in [36, 18, 12, 9, 6, 4, 3, 2, 1]
+    for flop in [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]])
+class ErtBroadwellTest(ErtTestBase):
+    def __init__(self, mpitask, flop):
```
Comment (vkarak) [inline cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py]: `mpitask` is a bit confusing as a name. I'd suggest to use either `num_tasks` or `num_ranks`.

File: cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+            'OMP_NUM_THREADS': str(self.num_cpus_per_task)
+        }
+        # Reference roofline boundaries for Intel Broadwell CPU (E5-2695 v4):
+        GFLOPs = 945.0
+        L1bw = 1788.0
+        L2bw = 855.0
+        L3bw = 547.0
+        DRAMbw = 70.5
+        # slowest job:
+        mpitaskm1 = 1
```
Comment (vkarak) [inline cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py]: Why not `mpitask_min` or `min_tasks`? Similarly below.

File: cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        flopm1 = 1024
+        self.roofline_rpt = 'rpt'
+        if mpitask == mpitaskm1 and flop == flopm1:
+            self.post_run = [
+                'cat *_job.out | python2 preprocess.py > pre',
+                'python2 maximum.py < pre > max',
+                'python2 summary.py < max > sum',
+                # give enough time for all the dependent jobs to collect data
+                'sleep 60',
+                'cat ../ert_FLOPS*/sum |python2 roofline.py > rpt',
```
Comment (vkarak) [inline cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py]: Would it make sense to have these Python scripts (`minimum.py`, `maximum.py`, `roofline.py`) inside this repo and not as resources?

File: cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+            ]
+            self.sanity_patterns = sn.all([
+                # --- check data type:
+                sn.assert_eq(sn.extractsingle(
+                    r'^\s+(?P<prec>\w+) \* __restrict__ buf = \(\w+ \*\)'
+                    r'malloc\(PSIZE\);', 'driver1.c', 'prec'), 'double'),
```
Comment (vkarak) [inline cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py]: Are you missing an argument to `assert_eq` here or is there sth wrong with the alignment?

File: cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py (hunk 9 above)
Comment (jgphpc) [reply cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py]: To distribute the code, we should verify this first: https://bitbucket.org/berkeleylab/cs-roofline-toolkit/raw/e7726010d2653ebc48ddced21c234980d3e25854/Copyright
We will know more after cug.

Comment (vkarak) [reply cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py]: I see, thanks.

**Comment by vkarak** [comment]
@teojgo Can you approve this PR?

File: cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+        flop_min = 1024
+        self.roofline_rpt = 'rpt'
+        if num_ranks == num_ranks_min and flop == flop_min:
+            self.post_run = [
+                'cat *_job.out | python2 preprocess.py > pre',
+                'python2 maximum.py < pre > max',
+                'python2 summary.py < max > sum',
+                # give enough time for all the dependent jobs to collect data
+                'sleep 60',
+                'cat ../ert_FLOPS*/sum |python2 roofline.py > rpt',
```
Comment (teojgo) [inline cscs-checks/tools/profiling_and_debugging/berkeleylab-roofline.py]: Just add a space after `|` from formatting consistency

**Comment by vkarak** [comment]
@jenkins-cscs retry none

**Pull request merged**